client/src/GameState.ts

```diff
@@ -115,32 +140,96 @@ export class GameState {
+        const board = this.board.map((row) => {
+            return row.map((el) => {
+                if (!el) {
+                    return el;
+                }
+                return el.toMonominoState();
+            });
+        });
+
+        const reverseRotations = (this.playerId || 0) % 4;
+        const boardFromPlayer0 = board.map((row) => row.slice());
+        for (let row = 0; row < BOARD_SIZE; row++) {
+            for (let col = 0; col < BOARD_SIZE; col++) {
+                const [newRow, newCol] = Tetromino.rotateCoords(
+                    [row, col],
+                    BOARD_SIZE,
+                    reverseRotations
+                );
+                boardFromPlayer0[newRow][newCol] = board[row][col];
+                const newBlock = boardFromPlayer0[newRow][newCol];
+                if (newBlock) {
+                    newBlock.position = [newRow, newCol];
+                }
+            }
+        }
+        return boardFromPlayer0;
+    }
+
+    public fromBoardState(boardState: BoardState): Array<Monomino> {
+        const needRedraw = [];
+        if (this.playerId && this.playerId > 0) {
+            // rotate remote player0 board to local board
+            const ccRotations = 4 - ((this.playerId || 0) % 4);
+            const localBoard = boardState.map((row) => row.slice());
+            for (let row = 0; row < BOARD_SIZE; row++) {
+                for (let col = 0; col < BOARD_SIZE; col++) {
+                    const [newRow, newCol] = Tetromino.rotateCoords(
+                        [row, col],
+                        BOARD_SIZE,
+                        ccRotations
+                    );
+                    localBoard[newRow][newCol] = boardState[row][col];
+                    const monominoState = boardState[row][col];
+                    if (monominoState) {
+                        monominoState.position = [newRow, newCol];
+                    }
+                }
+            }
+            boardState = localBoard;
+        }
+        for (let row = 0; row < BOARD_SIZE; row++) {
+            for (let col = 0; col < BOARD_SIZE; col++) {
+                const monominoState = boardState[row][col];
+                if (!monominoState) {
+                    // erase existing monomino if it should now be null
+                    const existingMonomino = this.board[row][col];
+                    if (existingMonomino) {
+                        existingMonomino.destroy();
+                    }
+                    this.board[row][col] = monominoState;
+                } else {
+                    const [monomino, shouldRedraw] =
+                        Monomino.upcreateFromMonominoState(
+                            monominoState,
+                            this.board[row][col]
+                        );
+                    this.board[row][col] = monomino;
+                    if (shouldRedraw) {
+                        needRedraw.push(monomino);
+                    }
+                }
```

Comment:
I feel like this might be cleaner if you combined the two double for loops, but maybe that's just code golf-y.